src/shims/env.rs

```diff
@@ -104,11 +103,52 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
         if let Some(old) = success {
             if let Some(var) = old {
-                this.memory_mut().deallocate(var, None, MiriMemoryKind::Env.into())?;
+                this.memory_mut()
+                    .deallocate(var, None, MiriMemoryKind::Env.into())?;
             }
             Ok(0)
         } else {
             Ok(-1)
         }
     }
+
+    fn getcwd(
+        &mut self,
+        buf_op: OpTy<'tcx, Tag>,
+        size_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`getcwd` not available when isolation is enabled")
+        }
+
+        let tcx = &{ this.tcx.tcx };
+
+        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
+        let size = this.read_scalar(size_op)?.to_usize(&*this.tcx)?;
+        // If we cannot get the current directory, we return null
+        match env::current_dir() {
+            Ok(cwd) => {
+                // It is not clear what happens with non-utf8 paths here
```

Comment:
Yeah this seems suspicious. And probably *very* wrong on Windows.

I think what we should do is have a helper function to put an `OsStr` into memory. This should use `OsStrExt`: on Unix, that has an `as_bytes` method that we can use to get the raw bytes. On Windows, we can get an iterator over `u16` that we can write to memory one after the other.

Reply:
Hm... but then we still have a problem in cross-execution. We have to be careful not to mix up the host `OsStr` and the target `OsStr`.

For now, it seems reasonable to add a check that the host and target OS are the same.

Reply:
Given that these shims only work for linux (and similar) right now I could write the function to get `OsStr` into miri and leave a comment saying that this needs to be taken care of when implementing the `fs` shims for windows.

Reply:
Well now you already implemented the `FileHandler` thing, might as well keep it... but we don't test cross-execution between Unix and non-Unix so I'd be surprised if it actually works. ;)

But then there should be a comment at the file handle table explaining that this is its purpose. It should also explain why we need to keep track of the flags.

> Given that these shims only work for linux (and similar) right now I could write the function to get OsStr into miri and leave a comment saying that this needs to be taken care of when implementing the fs shims for windows.

Sure, as long as the only place that would need fixing here is the helper function that seems fine.

Reply:
https://github.com/rust-lang/miri/pull/993